### DISCUSSION TITLE
Applcation crash

### DISCUSSION THREAD

**Discussion by tuxroger** [gabime/spdlog#2691] [Q&A]
Good afternoon,
I have been using spdlog for a while and I have faced many weird situations but this one is blowing my mind. I do not know what can be happenning.

This is the code snippet:

// Session Logging functions
bool PBConnectionThread::createSessionLogger(const std::string& loggerName, const std::string& filename)
{
	SPDLOG_LOGGER_INFO(logger, "");
	SPDLOG_LOGGER_INFO(logger, "Creating session log with filename '{}'.", filename);

	std::vector<spdlog::sink_ptr> sinks;

	auto rotating = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(filename, 1024 * 1024 * 50, 5, true);	// 50 MB
	if (rotating == nullptr)
	{
		SPDLOG_LOGGER_ERROR(logger, "Error while creating rotating_file_sink_mt file sink");
		return false;	// early exit
	}
	sinks.push_back(rotating);

	sessionLogger = std::make_shared<spdlog::logger>(loggerName, begin(sinks), end(sinks));
	if (sessionLogger == nullptr)
	{
		SPDLOG_LOGGER_ERROR(logger, "Error while creating spdlog logger");
		return false; // early exit
	}
	sessionLogger->set_level(spdlog::level::trace);

	spdlog::register_logger(sessionLogger);
	SPDLOG_LOGGER_INFO(logger, "Succesfully created connection logger!", filename);

	return true;
}

This code, from time to time is making my application crash. Here is an example of the output when it is crashing:

[2023-03-29 09:49:02.012] [ProbeBuilderServer] [info] PBConnectionThread::createSessionLogsDirectory[t3147436335914322354]
[2023-03-29 09:49:02.012] [ProbeBuilderServer] [info] PBConnectionThread::createSessionLog[t3147436335914322354]
[2023-03-29 09:49:02.012] [ProbeBuilderServer] [info] PBConnectionThread::createConnectionLogger[t3147436335914322354]
[2023-03-29 09:49:02.013] [ProbeBuilderServer] [info] PBConnectionThread::createConnectionLogger[t3147436335914322354]: Creating session log with filename 'LOGS/log_Richardson_{b0d2be45-b578-4bf5-a260-0f13125a54eb}_29Mar2023094902AM.log'.

ORA-24550: signal received: Unhandled exception: Code=e06d7363 Flags=1

As you can see something happens between the beginning of the function and the end of it. As you can also see, I am checking for all the possible nullptrs along the function code.

As I mentioned above, this crash occurs from time to time and I do not why.

Thanks for your time!

**Comment by tt4g** [comment]
The `e06d7363` code represents a system error.
It may be a system problem and a crash dump should be investigated.

See: [debugging \- Decoding the parameters of a thrown C\+\+ exception \(0xE06D7363\) \- Stack Overflow](https://stackoverflow.com/questions/9095898/decoding-the-parameters-of-a-thrown-c-exception-0xe06d7363)

In most cases, it is a lack of permissions to access folders or files, or a lack of memory.

**Reply by tt4g** [reply]
> ORA-24550: signal received: Unhandled exception: Code=e06d7363 Flags=1

`ORA-24550` is an Oracle database error code.
I thought spdlog was not the cause of the crash?

**Reply by tuxroger** [reply]
You are right, this error is always shown but I am not really performing any operation in the database backend while this is failing. That's why I thought this exception is thrown as a result of the application crashing and oracle throws an error while freeing up resources.

**Comment by tuxroger** [comment]
Hello, I just got a crash dump of the application. Here is a screenshot in which you can see that dropping the logger raises an exception: 
![image](https://user-images.githubusercontent.com/1750020/233715015-d94640a6-3bf3-4d04-91ba-91e0672f6a6f.png)
Do you have any idea what can be raising the exception?
Thanks!

**Reply by tt4g** [reply]
The `loggerName` variable is an invalid reference.

Or `logger_map_mutex_` is broken because the `PGConnectionThread` destructor is called after the application exits.
This is probably caused by not calling `spdlog::shutdown()` before the application exit.
https://github.com/gabime/spdlog/blob/0ca574ae168820da0268b3ec7607ca7b33024d05/include/spdlog/details/registry-inl.h#L219-L227

**Reply by tuxroger** [reply]
loggerName is actually this:

dropSessionLogger(sessionLogger->name());

and sessionLogger is std::shared_ptr<spdlog::logger> sessionLogger; and it is not null since it would crash in some other way.

How can the name of a logger be an invalid reference?

I am calling shutdown at the end of the application just before exiting but this specific case was a crash and that's why you don't see the call in the stack.

**Reply by tt4g** [reply]
If the `PGConnectionThread` destructor is being called after the application exit, then `logger_map_mutex_` may be broken.
There is no guarantee that `std::mutex` will work after `main()` function.

**Reply by tuxroger** [reply]
The PBConnectionThread dtor is not being called after main(). I have a pool of threads and this is one of them being destroyed but I have more threads in the pool which are alive.
I mean, this crash did not happen when shutting down the application, it crashes while running. 
Quick question, is there any problem that the loggerName is something like this? "{edf71852-9b22-4ddc-88b7-dee185cc2687}". I understand it is not a problem since it is a string but...
Thanks for your support!
Roger

**Comment by gabime** [comment]
Just a guess but perhaps try to replace `dropSessionLogger(sessionLogger->name());`
with `dropSessionLogger(std::string(sessionLogger->name()));` 
Could be that the logger_name get destroyed in drop() and then used again in the `if (default_logger_ && default_logger_->name() == logger_name`

**Reply by tuxroger** [reply]
Ok thanks, I will try that. 
This crash happens from time to time so it might be the case that from time to time (race condition) the logger gets destroyed before.
I have also a variable containing the logger name. I will try to use that string instead of the logger itself. 
Thanks for your help!

**Comment by gabime** [comment]
@tuxroger Let us know if this fix this. Because if yes, spdlog's drop could be improved to avoid such cases.

**Reply by tuxroger** [reply]
Hello, 
I already call drop without calling logger->name(). As this error only happens from time to time, I will assume that after certain time without failing the issue might be resolved. I will let you know the outcome.
Thanks again for your support!

**Comment by tuxroger** [comment]
Hi there,
I just wanted to let you know that my application is crashing now when creating the session logger.
Please find attached the stack trace of the error.
Thanks in advance for your time!
![image](https://user-images.githubusercontent.com/1750020/235169783-5c1ec313-49d8-460a-abc2-db9404921d32.png)

**Reply by tt4g** [reply]
It looks like `throw` is being invoked, so `catch (const std::exception &ex)` and print `ex.what()`.

**Reply by tuxroger** [reply]
Yes, I will do so. Just for me to know, should I wrap around try / catch all the calls used to create / remove / configure loggers and sinks?
I assume info() debug() and trace() are not throwing exceptions right?
I was not catching any exception since I did not see anywhere in the doc that these calls shall raise exceptions.
I will keep you up to date, thanks!

**Reply by tt4g** [reply]
exception is what is thrown when a program does not meet the conditions for proper operation.
If you can check that the conditions for the program to work have been met in advance, there is no need for `try`-catch` protection.

For example, file sync will `throw` an exception when it does not have permission to create a file.
This should be confirmed before creating the file sink.

An example of a condition that cannot be confirmed in advance is a pattern where the system to be connected to, such as a connection to a database, may be down.

Next, if your application will not work when the code enclosed in `try`-`catch` is not executed, your application must abort after printing the error message in `catch`.

The developer must consider whether to `try`-`catch` and what to do with `catch`.

**Reply by tuxroger** [reply]
First of all, thanks for the explanation. I know what exceptions are, by the way. What I wanted to know is if in your documentation is there anywhere specifying which calls can throw. Like when you specify in the signature of a method the keyword "throws" so that anybody using that needs to be aware and try/catch that call.
I understand that when working with files, for example, many thing can happen (buffer underrun, write permission issues, etc). Bu depending on the implementation of your API, you could catch exceptions and return false.
If that is not the case, I just wanted to know where should I be careful when using your api.
Thanks a lot for your time!

**Reply by tt4g** [reply]
You can find the throw rule of exception on the Wiki.
https://github.com/gabime/spdlog/wiki/Error-handling

> Spdlog will not throw exceptions while logging (since version 39cdd08).

The constructor is not a logging API.

**Reply by tuxroger** [reply]
Thanks a lot! that makes a lot of sense, you don't want exceptions while logging. Thanks a lot! I will try to modify my code so when construction loggers I try/catch the block.
Thanks!

**Reply by tuxroger** [reply]
Good afternoon.
The application has crashed twice and I have the what() message for every exception that has been catched:

1.
rotating_file_sink: failed renaming C:\Users\probebuilder\AppData\Local\PBServer\LOGS\log_Richardson_28Apr2023154849PM.log to C:\Users\probebuilder\AppData\Local\PBServer\LOGS\log_Richardson_28Apr2023154849PM.1.log: permission denied

2.
rotating_file_sink: failed renaming C:\Users\probebuilder\AppData\Local\PBServer\LOGS\log_Richardson_28Apr2023160059PM.log to C:\Users\probebuilder\AppData\Local\PBServer\LOGS\log_Richardson_28Apr2023160059PM.1.log: permission denied

It looks like when changing the current .log file to the next one with the .1 suffix, there are problems with the permission / ownership of the current file (.log).

This has never happened to me

Thanks for your time!

**Reply by tuxroger** [reply]
By the way, according to the documentation only in the construction and destruction exceptions might be thrown but what happens if you try to log in a logger that is null?

SPDLOG_LOGGER_INFO(logger, "blahblah");

It simply does not log? Or a previous check should be performed so I dont access nullptr?

Thanks!

**Reply by tt4g** [reply]
The "permission denied" is a security problem and is not something spdlog has any control over.
Please make sure that you manage the machine's permissions yourselves.

The `nullptr` check should be done by you.

**Reply by tuxroger** [reply]
Thanks for the answer!
It is weird that the process has not privileges since it is the same proces that created the .log file and when it is full it is unable to rename it to .1.log
Do you have an idea how can this happen? I understand that if the file is in use by another process we could not rename it but the only one with the file open is the process that just created the file...
I will also try to investigate on my own and let you know.
